### PR TITLE
CUDA-based solver class and autograd function

### PR DESCRIPTION
**Pull request by maurimo**
This PR adds solver and autograd function based on CUDA solver and operations.

Since CUDA "refactor" library has the limitation of sharing an opaque context for the symbolic decomposition and the numerical factor data, a number of context is used in order to allow (a limited amount of) optimization loop unrolling. If a context was overwritten when calling backward step an exception is raised.

(TODO: add here bits of previous discussion)

### CODE REVIEW

**Comment by maurimo** [comment]
Set parent branch as target.
Added damping parameters, now the solver supports the same damping types as the dense solver (ellipsoidal or not), the autograd is tested using "large" damping parameters (a correction in the autograd function may actually be overkill, and needed only in the test with very large values).

File: theseus/optimizer/linear/lu_cuda_sparse_solver.py
Comment (luisenp) [inline theseus/optimizer/linear/lu_cuda_sparse_solver.py:85]: Would it be safer to completely error out instead of doing FIFO? If I understand this correctly, if this solver is used more than `_num_solver_contexts`, the backward pass will start giving incorrect results. Or maybe this is already caught [here](https://github.com/facebookresearch/theseus/pull/24/files#diff-69f2100fbaed4dc21b13b12f5990f5b34eeb26adfec87de4a45385a21b0aa061R110)? Even if that's the case, is there a reason we would like to keep things mod `_num_solver_context` in this function?

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (luisenp) [inline theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:27]: Are we testing unrolling operations? Even w/o the implicit differentiation checks, we could add some basic test for this werein we do a series of `solve()` operations and check some tensor's `.grad` content numerically?

**Review by luisenp — COMMENTED** [review]
Looks great! Left a couple of questions/comments for discussion.

File: theseus/__init__.py
Comment (mhmukadam) [inline theseus/__init__.py:49]: maybe just do `from torch.cuda import is_available`?

File: theseus/__init__.py
Comment (mhmukadam) [inline theseus/__init__.py:51]: Should we have an else here that raises an error/warning, if the user tries to call/use `LUCudaSparseSolver` without having cuda.

File: theseus/optimizer/autograd/__init__.py
Comment (mhmukadam) [inline theseus/optimizer/autograd/__init__.py:15]: Same comment about torch import and use without cuda.

File: theseus/optimizer/autograd/tests/test_lu_cuda_sparse_backward.py
Comment (mhmukadam) [inline theseus/optimizer/autograd/tests/test_lu_cuda_sparse_backward.py:12]: Following convention from all other test files we should replace with `import theseus as th`.

File: theseus/optimizer/linear/__init__.py
Comment (mhmukadam) [inline theseus/optimizer/linear/__init__.py:13]: Flagging here too so we remember to make it consistent with what we end up doing in other places.

File: theseus/optimizer/linear/lu_cuda_sparse_solver.py
Comment (mhmukadam) [inline theseus/optimizer/linear/lu_cuda_sparse_solver.py:18]: We should move this to `linear/sparse_solver.py` so change that file to be Cholmod sparse solver since that's the only one in there currently. Same for the test file.

File: theseus/optimizer/linear/lu_cuda_sparse_solver.py
Comment (mhmukadam) [inline theseus/optimizer/linear/lu_cuda_sparse_solver.py:43]: Is LM not supported at the moment?

File: theseus/optimizer/autograd/lu_cuda_sparse_autograd.py
Comment (mhmukadam) [inline theseus/optimizer/autograd/lu_cuda_sparse_autograd.py:115]: We don't have to do it in this PR, but aside from a few lines like this one, we might be able to pull in a lot of common code between LU sparse and Cholmod sparse.

This could also work for the solver class itself by having a class hierarchy
`LinearSolver` -> `SparseSolver` ->  `CholmodSparseSolver` and `LUCudaSparseSolver`

cc @luisenp

**Review by mhmukadam — COMMENTED** [review]
Looks great! Only some minor comments and one comment about whether we should consolidate some of the contents in the sparse solvers with a common sparse solver parent class.

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (mhmukadam) [reply theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:27]: +1

File: theseus/__init__.py
Comment (luisenp) [reply theseus/__init__.py:51]: If cuda is not available, then `LUCudaSparseSolver` would not even be visible, so trying to do `import th.LUCudaSparseSolver` would throw an error. Maybe this will be confusing for users?

Is it possible to make all of these imports permanent in the relevant `__init__.py` files, but address the `cuda.is_available()` issue inside the corresponding `.py` files? Basically, allow users to import `LUCudaSparseSolver` and only throw an error if this class is attempted to be instantiated w/o cuda being available.

Comment (mhmukadam) [reply theseus/__init__.py:51]: Throwing an error seems okay but your other suggestion is also good. Not sure if there is a standard convention here that we can follow.

Comment (luisenp) [reply theseus/__init__.py:51]: Right, but the error will be something like:

```python
import th
solver = th.LUCudaSparseSolver(...)
...
...
AttributeError: module 'theseus' has no attribute 'LUCudaSparseSolver'
```

which seems misleading, no?

Comment (luisenp) [reply theseus/__init__.py:51]: What we want is for this to be a `RuntimeError` that says `"class `LUCudaSparseSolver` requires cuda to be available"`. This could be achieved, for example, by adding the check inside the class' constructor, or, if that's not possible due to some C++ import issues, making a stub class with no requirements whose job is only to throw an error, and that gets imported in place of the original one when no cuda is present.

Comment (mhmukadam) [reply theseus/__init__.py:51]: Yes, we would want an error message that indicates this functionality is only available with cuda. Whichever version get us there easily sounds good.

File: theseus/optimizer/linear/lu_cuda_sparse_solver.py
Comment (maurimo) [reply theseus/optimizer/linear/lu_cuda_sparse_solver.py:85]: so, yes, there is where the error is caught. The mod-_num_solver_contexts logic is not here but in the Sparse solver, that instead of giving this class a sparse representation is providing this class a "new" context (new = least recently used, "mod" num contexts). Here (where you linked) it's just necessary to check that the context we have been provided has not been reused in the mean time, can do so comparing the "factor_id". If you didn't notice where that is done I think it was indeed because this fact was quite non-intrusive in the overall logic ;)

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (maurimo) [reply theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:27]: that can certainly be done!

File: theseus/__init__.py
Comment (maurimo) [reply theseus/__init__.py:51]: ok but at the level of the python class right? so no problems creating a mock python class which just throws error then I think? This may confuse python typing unless the mock class is mimicking all the methods with the same types.

File: theseus/optimizer/linear/lu_cuda_sparse_solver.py
Comment (maurimo) [reply theseus/optimizer/linear/lu_cuda_sparse_solver.py:18]: I would prefer keeping files separate, it looks cleaner (even more so if the cuda class will have a mock version when cuda is not available). Anyway all is re-exported in the `__init__` file so it makes no difference to the user right?

File: theseus/optimizer/linear/lu_cuda_sparse_solver.py
Comment (maurimo) [reply theseus/optimizer/linear/lu_cuda_sparse_solver.py:43]: dead code, the parameter is set in solve, will be cleaned.

File: theseus/optimizer/autograd/lu_cuda_sparse_autograd.py
Comment (maurimo) [reply theseus/optimizer/autograd/lu_cuda_sparse_autograd.py:115]: Talking about autograd function or the solver itself? the solver classes look to small to find any relevant common code... the autograd function I would say yes, even if the way to merge some functionality will certainly require some thoughts.

File: theseus/__init__.py
Comment (luisenp) [reply theseus/__init__.py:51]: Oh, but that's a good point about typing and the mock class. Is there a reason why the error cannot be thrown inside the original class' constructor? Can you elaborate a bit?

Comment (maurimo) [reply theseus/__init__.py:51]: so, the problem is that if cuda is not available I cannot import any C++ extension depending on cuda either (or an error occurs). Therefore the "import" has to be done inside the functions themselves. Do you see any issue with that?

File: theseus/optimizer/autograd/lu_cuda_sparse_autograd.py
Comment (mhmukadam) [reply theseus/optimizer/autograd/lu_cuda_sparse_autograd.py:115]: Autograd function does have a lot of shared code, but I think it's worth considering the solver too.

File: theseus/optimizer/linear/lu_cuda_sparse_solver.py
Comment (mhmukadam) [reply theseus/optimizer/linear/lu_cuda_sparse_solver.py:18]: That sounds fine too. Then, after we decide what to do with the class hierarchy (from the other comment below) for the solver we should either rename sparse solver to cholmod sparse or make a new cholmod sparse with only the parent class in sparse solver to make things consistent.

File: theseus/optimizer/autograd/lu_cuda_sparse_autograd.py
Comment (luisenp) [inline theseus/optimizer/autograd/lu_cuda_sparse_autograd.py:14]: Any reason we cannot move all of these conditional imports to the top of the file?

File: theseus/optimizer/autograd/lu_cuda_sparse_autograd.py
Comment (luisenp) [inline theseus/optimizer/autograd/lu_cuda_sparse_autograd.py:119]: Same question about moving to the top of the file.

File: theseus/optimizer/linear/lu_cuda_sparse_solver.py
Comment (luisenp) [inline theseus/optimizer/linear/lu_cuda_sparse_solver.py:26]: Same question about moving to top of the file.

File: theseus/optimizer/linear/lu_cuda_sparse_solver.py
Comment (luisenp) [inline theseus/optimizer/linear/lu_cuda_sparse_solver.py:45]: Same question about moving to top of the file.

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (luisenp) [inline theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:110]: dead code?

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (luisenp) [inline theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:6]: I tried to run this in my computer, before I installed the CUDA extensions, and got an error as shown below. Do we want to make the tests in this file conditional on the extension being installed? We could use [pytest.importskip](https://docs.pytest.org/en/6.2.x/reference.html#pytest-importorskip) for this. 

image

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (luisenp) [inline theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:132]: I was a bit worried about the high error tolerance, but changing epsilon to `1e-7` and both of the tolerances to `1e-4` works for me. Maybe better?

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (luisenp) [inline theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:140]: It would be useful to add some brief explanatory comments here in case someone else needs to look at this test in the future. It seems that this test checks against a directional derivative (in the direction of `perturb`), is this correct? Beforehand, I was looking for something like a partial derivative wrt to each of A's dimensions. Took me a bit to figure out what this test was doing.

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (luisenp) [inline theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:107]: Would be helpful to add some comments with a high level summary of what this does. If I understand correctly, this returns

```
y = dot(z,c)
z = sum(A_i \ b_i)
```

**Review by luisenp — CHANGES_REQUESTED** [review]
Looks great! Left a bunch of minor comments, but they should all be quick to fix. After that merging is OK with me.

File: theseus/optimizer/autograd/lu_cuda_sparse_autograd.py
Comment (maurimo) [reply theseus/optimizer/autograd/lu_cuda_sparse_autograd.py:14]: Well if so we are back to having to implement a mock-class for when cuda is not available...

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (maurimo) [reply theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:110]: removing in next rev!

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (maurimo) [reply theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:6]: well that's precisely what the ci currently does, they are only tested when the extension is installed.
However I would not make them run only if the extension is *available*, but rather on where the user tried to install with Cuda. In such a case they should fail and notify the user that something didn't work as expected and the extension she tried to compile is actually not available.

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (maurimo) [reply theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:132]: makes sense, will change those in next rev, thanks for trying!
Anyway the check is not trivial with current values (eg as could be because if values verified were very close to zero), I had verified that too dumping the content of the derivatives.

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (maurimo) [reply theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:140]: makes sense! I thought the fastest way to get a test working with with a random directional check (on several directions being sampled).

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (maurimo) [reply theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:107]: will add comments yeah

File: theseus/optimizer/autograd/lu_cuda_sparse_autograd.py
Comment (luisenp) [reply theseus/optimizer/autograd/lu_cuda_sparse_autograd.py:14]: I'm mostly wondering about any overheads from repeated `import` calls. I know python caches the module for future imports, but there might some additional non-negligible cost for calling `import`? (not sure). We could get around writing a mock class by doing the conditional import at the top while also setting a boolean flag, say `_EXTLIB_WAS_LOADED`, and (inside the class) throw the runtime error based on the status of the flag (rather than  checking the import).

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (luisenp) [reply theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:140]: Just to be clear, I think the test makes a lot of sense, just mentioned the alternative to explain why it took me a bit to figure this out, and thus why a comment in the code would be helpful.

File: theseus/optimizer/autograd/lu_cuda_sparse_autograd.py
Comment (maurimo) [reply theseus/optimizer/autograd/lu_cuda_sparse_autograd.py:14]: I see - yeah that would work. I would go for the cleaner solution, just not sure which is (ideas @mhmukadam ?) if there is an overhead doing import it will be certainly very small, and I feel we should not optimize around something that should already be optimized inside python.

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (maurimo) [reply theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:6]: So, that's right that running tests should not fail because theseus was correctly installed with no cuda support!
Not sure how that can be achieved, I was just saying that the test should just not skip because if was unable to load the extension. Ideally there should be a command running tests that will skip cuda tests if the extension is not installed, or two separate commands for the user to check both scenarios.
I think we have to minimize the confusion for the user, considering the two scenarios: 1. the user willingly disabled cuda extensions 2. the user wants cuda extensions, but they may not have installed properly, or be nonfunctional, and tests should check that. Suggestions here too @mhmukadam ?

File: theseus/optimizer/autograd/lu_cuda_sparse_autograd.py
Comment (luisenp) [reply theseus/optimizer/autograd/lu_cuda_sparse_autograd.py:14]: I checked with other engineers, and importing does have some overhead, but perhaps it's negligible. Since we are doing benchmarking this half anyway, I vote for keeping this version, and revisit the issue if we see opportunities for improvement later. Sounds good?

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (luisenp) [reply theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:6]: As discussed offline, let's make this test optional on depending on the extension being installed, and consider other mechanisms (e.g., a separate command) for testing that the installation was successful. Additionally, `pytest.importskip` has a `reason=` kwarg that we can use to print a warning message to alert those users that did in fact want to run the CUDA test that their install might be incorrect.

File: theseus/optimizer/autograd/lu_cuda_sparse_autograd.py
Comment (mhmukadam) [reply theseus/optimizer/autograd/lu_cuda_sparse_autograd.py:14]: Sounds good. I would have proposed the same thing -- keep this and during benchmarking see if this is a bottle in performance to be prioritize and fixed.

**Review by mhmukadam — COMMENTED** [review]
Looks good. We should be able to merge once the pending comments are resolved in the final rev.

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (maurimo) [reply theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:6]: in new revision I'm adding the "cuda" mark to cuda tests, and a note on the readme on how to skip those tests

**Comment by maurimo** [comment]
this new rev should address all the concerns in the comment, please give a look! @luisenp @mhmukadam

File: theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py
Comment (luisenp) [reply theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py:6]: This is great!

**Comment by luisenp** [comment]
LGTM! Unless @mhmukadam has some more comments, I'm OK with merging now.

**Review by mhmukadam — APPROVED** [review]
Looks great, thanks for the final push!
(Flagged an issue for the consolidation of some of the autograd code which we can handle in a future PR)

**Pull request merged**
